tools/generateDeclarationFile.ts

```diff
@@ -0,0 +1,286 @@
+import {
+  Diagnostic,
+  ExportedDeclarations,
+  IndentationText,
+  ModuleDeclarationStructure,
+  Node,
+  Project,
+  SourceFile,
+  Statement,
+  StatementStructures,
+  Structure,
+  StructureKind,
+  SyntaxKind,
+  WriterFunction,
+} from "./deps.ts";
+
+console.log("Generating declaration file...");
+const statements: (StatementStructures | WriterFunction)[] = [];
+const declarationProject = getDeclarationProject();
+
+const indexFile = declarationProject.getSourceFileOrThrow(
+  "./dist/index.d.ts",
+);
+
+// header
+statements.push((writer) => {
+  writer
+    .writeLine(
+      `// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.`,
+    )
+    .blankLine()
+    .writeLine(`/// <reference types="node" />`)
+    .blankLine()
+    .writeLine(`import { URL } from "url";`)
+    .writeLine(`import * as undici from "undici";`)
+    .blankLine();
+});
+
+statements.push(...getMainStatements());
+
+statements.push({
+  kind: StructureKind.Module,
+  name: `"deno.ns/deno"`,
+  statements: (writer) => {
+    writer.writeLine("export = Deno;");
+  },
+});
+
+statements.push({
+  kind: StructureKind.Module,
+  name: `"deno.ns/global"`,
+});
+
+statements.push({
+  kind: StructureKind.Module,
+  name: `"deno.ns/test-internals"`,
+  statements: [
+    ...Array.from(
+      fileExportsToStructures(declarationProject.getSourceFileOrThrow(
+        "./dist/deno/internal/test.d.ts",
+      )),
+    ).map((s) => exportAndStripAmbient(s)),
+  ],
+});
+
+// Create a new project with limited declarations and add the declaration
+// file to it. Save the file then type check.
+const newProject = new Project({
+  compilerOptions: {
+    // limit to only node types
+    types: ["node"],
+  },
+  tsConfigFilePath: "./tsconfig.json",
+  skipAddingFilesFromTsConfig: true,
+  manipulationSettings: {
+    indentationText: IndentationText.TwoSpaces,
+  },
+});
+const sourceFile = newProject.createSourceFile(
+  "./lib/deno.ns.lib.d.ts",
+  { statements },
+  { overwrite: true },
+);
+
+sourceFile.saveSync();
+
+exitIfDiagnostics(newProject, sourceFile.getPreEmitDiagnostics());

```

Comment:
Diagnostics are checked for here and above it limits it to only "node" types. This ensures we don't distribute a declaration file that has issues.